### PR TITLE
Integrate subdomain filter

### PR DESCRIPTION
**Pull request by MatMoore**
When selecting a domain, a user may select a top-level domain and a subdomain.

When the subdomain is selected, `subdomain` is passed as an additional form value.

If this is present, we need to:

- Use this as the sole domain filter
- Preserve the selected value
- Add the subdomain to the label shown in "Selected filters"
- Exclude the subdomain field from the clear link used in "Selected filters"

Please review carefully - the logic is very awkward.

### CODE REVIEW

File: home/forms/domain_model.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+from typing import NamedTuple
```
Comment (MatMoore) [inline home/forms/domain_model.py:1]: This information should eventually come from the catalogue. I decided against implementing that now, but I thought it was worth extracting out  into a separate module, as there are a lot of places we use this metadata

Comment (MatMoore) [reply home/forms/domain_model.py:1]: One benefit of pulling this from the catalogue is we can include counts of the number of assets in each domain, which would assist users browsing the catalogue just to get a sense of what's there.

File: tests/selenium/test_search_scenarios.py [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+from tests.conftest import generate_page, mock_search_response
 
 from .helpers import check_for_accessibility_issues
-from tests.conftest import mock_search_response, generate_page
-from data_platform_catalogue.search_types import ResultType
 
 
 @pytest.mark.slow
-class TestSearchWithoutJavascriptAndCss:
+class TestSearch:
```
Comment (MatMoore) [inline tests/selenium/test_search_scenarios.py:12]: Turns out this does actually include assets. The `live_server` fixture decides whether to include assets based on the presence of `django.contrib.staticfiles` in `INSTALLED_APPS`.

File: home/forms/domain_model.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class Domain(NamedTuple):
+    value: str
```
Comment (tom-webber) [inline home/forms/domain_model.py:5]:
```suggestion
    urn: str
```
small suggestion just to make the code a little easier to parse (propagated)

File: home/forms/domain_model.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        self.top_level_domains = [
+            Domain("urn:li:domain:HMCTS", "HMCTS"),
+            Domain("urn:li:domain:HMPPS", "HMPPS"),
+            Domain("urn:li:domain:HQ", "HQ"),
+            Domain("urn:li:domain:LAA", "LAA"),
+            Domain("urn:li:domain:OPG", "OPG"),
+        ]
+
+        for value, label in self.top_level_domains:
+            self.labels[value] = label
```
Comment (tom-webber) [inline home/forms/domain_model.py:26]:
```suggestion
        for urn, label in self.top_level_domains:
            self.labels[urn] = label
```

File: home/forms/domain_model.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                    "urn:li:domain:5c423c06-d328-431f-8634-7a7e86928819",
+                    "Public defender",
+                ),
+            ],
+        }
+
+        for domain, subdomains in self.subdomains.items():
+            domain_label = self.labels[domain]
+            for value, label in subdomains:
+                self.labels[value] = f"{domain_label} - {label}"
```
Comment (tom-webber) [inline home/forms/domain_model.py:85]:
```suggestion
            for urn, subdoman_label in subdomains:
                self.labels[urn] = f"{domain_label} - {subdoman_label}"
```

File: home/forms/domain_model.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        subdomains = []
+        for domain_choices in self.subdomains.values():
+            subdomains.extend(domain_choices)
+        return subdomains
+
+    def get_parent_value(self, subdomain_value) -> str | None:
+        for domain, subdomains in self.subdomains.items():
+            for s in subdomains:
+                if subdomain_value == s[0]:
+                    return domain
```
Comment (tom-webber) [inline home/forms/domain_model.py:100]:
```suggestion
    def get_parent_urn(self, child_subdomain_urn) -> str | None:
        for domain, subdomains in self.subdomains.items():
            for subdomain in subdomains:
                if child_subdomain_urn == subdomain.urn:
                    return domain
```

File: home/forms/domain_model.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+    def get_label(self, value):
+        return self.labels.get(value, value)
```
Comment (tom-webber) [inline home/forms/domain_model.py:103]:
```suggestion
    def get_label(self, urn):
        return self.labels.get(value, urn)
```

File: home/forms/search.py [hunk 8]
```diff
@@ -3,63 +3,22 @@
...
-    # domain_list = facets.options("domains")
-    return [
-        ("", "All Domains"),
-        ("urn:li:domain:HMCTS", "HMCTS"),
-        ("urn:li:domain:HMPPS", "HMPPS"),
-        ("urn:li:domain:HQ", "HQ"),
-        ("urn:li:domain:LAA", "LAA"),
-        ("urn:li:domain:OPG", "OPG"),
+    choices: list[tuple[str, str]] = [
+        ("", "All domains"),
```
Comment (tom-webber) [inline home/forms/search.py:12]:
```suggestion
def get_domain_choices() -> list[Domain]:
    """Make API call to obtain domain choices"""
    choices: list[Domain] = [
        Domain("", "All domains"),
```
The other elements of this list are `Domain`s, so bringing this in line and increasing specificity of the typing

File: home/forms/search.py [hunk 9]
```diff
@@ -3,63 +3,22 @@
...
-            (
-                "urn:li:domain:24344488-d770-437a-ba6f-e6129203b927",
-                "Civil legal advice",
-            ),
-            ("urn:li:domain:Legal%20Aid", "Legal aid"),
-            ("urn:li:domain:5c423c06-d328-431f-8634-7a7e86928819", "Public defender"),
-        ],
-    }.get(domain, [])
+def get_subdomain_choices():
+    choices: list[tuple[str, str]] = [("", "All subdomains")]
```
Comment (tom-webber) [inline home/forms/search.py:19]:
```suggestion
def get_subdomain_choices() -> list[Domain]:
    choices: list[Domain] = [Domain("", "All subdomains")]
```
(same as above)

File: home/forms/search.py [hunk 10]
```diff
@@ -82,6 +41,25 @@ def get_where_to_access_choices():
...
+
+    def create_option(
+        self, name, value, label, selected, index, subindex=None, attrs=None
+    ):
+        option = super().create_option(
+            name, value, label, selected, index, subindex, attrs
+        )
+
+        if value:
+            option["attrs"]["data-parent"] = self.domain_model.get_parent_urn(value)
```
Comment (tom-webber) [inline home/forms/search.py:58]:
```suggestion
    def create_option(
        self, name, urn, label, selected, index, subindex=None, attrs=None
    ):
        option = super().create_option(
            name, urn, label, selected, index, subindex, attrs
        )

        if urn:
            option["attrs"]["data-parent"] = self.domain_model.get_parent_urn(urn)
```

**Review by mitchdawson1982 — APPROVED** [review]
LGTM

**Pull request merged**